app/controllers/invoice_controller.rb

```diff
@@ -0,0 +1,43 @@
+require 'net/http'
+require 'uri'
+class InvoiceController < ApplicationController
+  def index
+  end
+
+  def show
+    invoice_db = InvoiceDB.new
+    service = RetreiveClientService.new(invoice_db)
```

Comment:
Oh, i mean, i understand 'indirection' as 'unnecessary variable assignment'. i could put:
service = RetreiveClientService.new(InvoiceDB.new), looks a little bit more clean.

So your question was probably about why i pass DB object to Service like this. And the answer why i do it is that i like to keep logic separated from other layers of the system ( that is why i user service objects ), and in this case additional benefit also is that during the tests, if the system grows bigger, i can swap this DB service with InMemory one and this will make them a lot faster. Nobody likes to wait for the tests ;)